src/main/java/org/assertj/core/api/AssertionErrorCollectorImpl.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2012-2020 the original author or authors.
+ */
+package org.assertj.core.api;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public class AssertionErrorCollectorImpl implements AssertionErrorCollector {
+
+  volatile boolean wasSuccess = true;
+  List<AssertionError> assertionsCollected = Collections.synchronizedList(new ArrayList<>());
+  
+  AfterAssertionErrorCollected callback = this;
+
+  AssertionErrorCollector delegate = null;
+  
+  public AssertionErrorCollectorImpl() {
+    super();
+  }
+
+  // I think ideally, this would be set in the constructor and made final;
+  // however that would require a new constructor that would not make it
+  // backward compatible with existing SoftAssertionProvider implementations.
+  @Override
+  public void setDelegate(AssertionErrorCollector delegate) {
+    this.delegate = delegate;
+  }
+  
+  @Override
+  public Optional<AssertionErrorCollector> getDelegate() {
+    return Optional.ofNullable(delegate);
+  }
+  
+  @Override
+  public void collectAssertionError(AssertionError error) {
+    if (delegate == null) {
+      assertionsCollected.add(error);
+      wasSuccess = false;
+    } else {
+      delegate.collectAssertionError(error);
+    }
+    callback.onAssertionErrorCollected(error);
+  }
+
+  /**
+   * Returns a copy of list of soft assertions collected errors.
+   * @return a copy of list of soft assertions collected errors.
+   */
+  @Override
+  public List<AssertionError> assertionErrorsCollected() {
+    if (delegate == null) {
+      return Collections.unmodifiableList(assertionsCollected); 
+    }
+    return delegate.assertionErrorsCollected();
```

Comment:
If the delegate doesn't want clients to be messing with its internals, then it will return a defensive copy too. I want to avoid unnecessarily making a copy of a copy.

Reply:
If you feel strongly about it, I will wrap it even if the call is delegated. However, my feeling at the moment is that it should remain as it is.

Reply:
fair enough